handlers/viber/viber.go

```diff
@@ -10,3 +31,386 @@ POST /handlers/viber_public/uuid?sig=sig
+				if err != nil {
+					return nil, err
+				}
+				rr, err := utils.MakeHTTPRequest(req)
+				if err != nil {
+					return nil, err
+				}
+
+				viberMsgSize = rr.ContentLength
+
+			case "audio":
+				viberMessageType = "file"
+				viberMediaURL = mediaURL
+				req, err := http.NewRequest(http.MethodHead, mediaURL, strings.NewReader(""))
+				if err != nil {
+					return nil, err
+				}
+				rr, err := utils.MakeHTTPRequest(req)
+				if err != nil {
+					return nil, err
+				}
+				viberMsgSize = rr.ContentLength
+				filename = "Audio"
+
+			default:
+				status.AddLog(courier.NewChannelLog("Unknown media type: "+mediaType, msg.Channel(), msg.ID(), "", "", courier.NilStatusCode,
+					"", "", time.Duration(0), fmt.Errorf("unknown media type: %s", mediaType)))
+
+			}
+		}
+
+		viberMsg := viberOutgoingMessage{
+			AuthToken:    authToken,
+			Receiver:     msg.URN().Path(),
+			Text:         part,
+			Type:         viberMessageType,
+			TrackingData: msg.ID().String(),
+			Media:        viberMediaURL,
+			FileName:     filename,
+			Keyboard:     replies,
+		}
+
+		if viberMsgSize != "" {
+			viberMsgSizeInt, err := strconv.Atoi(viberMsgSize)
+			if err != nil {
+				return nil, err
+			}
+			viberMsg.Size = viberMsgSizeInt
+		}
+
+		requestBody := &bytes.Buffer{}
+		err := json.NewEncoder(requestBody).Encode(viberMsg)
+		if err != nil {
+			return nil, err
+		}
+
+		// build our request
+		req, err := http.NewRequest(http.MethodPost, sendURL, requestBody)
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Accept", "application/json")
+		rr, err := utils.MakeHTTPRequest(req)
+
+		// record log
+		log := courier.NewChannelLogFromRR("Message Sent", msg.Channel(), msg.ID(), rr)
+		status.AddLog(log)
+		if err != nil {
+			log.WithError("Message Send Error", err)
+			return status, nil
+		}
+
+		responseStatus, err := jsonparser.GetInt([]byte(rr.Body), "status")
+		if err != nil {
+			log.WithError("Message Send Error", errors.Errorf("received invalid JSON response"))
+			status.SetStatus(courier.MsgFailed)
+			return status, nil
+		}
+		if responseStatus != 0 {
+			log.WithError("Message Send Error", errors.Errorf("received non-0 status: '%d'", responseStatus))
+			status.SetStatus(courier.MsgFailed)
+			return status, nil
+		}
+
+		status.SetStatus(courier.MsgWired)
+		replies = ""
+	}
+	return status, nil
+}
+
+type viberOutgoingMessage struct {
+	AuthToken    string            `json:"auth_token"`
+	Receiver     string            `json:"receiver"`
+	Text         string            `json:"text"`
+	Type         string            `json:"type"`
+	TrackingData string            `json:"tracking_data"`
+	Sender       map[string]string `json:"sender,omitempty"`
+	Media        string            `json:"media,omitempty"`
+	Size         int               `json:"size,omitempty"`
+	FileName     string            `json:"file_name,omitempty"`
+	Keyboard     string            `json:"keyboard,omitempty"`
```

Comment:
This should be a pointer to `viberKeyboard`, not a string. Right now it is going to turn into a double encoded JSON blob

Reply:
that way the omitempty is not effective and there is keyboard: {} in the JSON

Reply:
You make it a pointer to a `viberKeyboard`, omit empty will leave it out then.

IE: 

`Keyboard *viberKeyboard`

Reply:
I have tried this and I keep getting {} in the JSON

Reply:
never mind, this is now updated